File: hard/print_check.lua
function div(a, b) return math.floor(a/b) end

s = {[0]="", "One", "Two", "Three", "Four",
     "Five", "Six", "Seven", "Eight", "Nine",
     "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen",
     "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen",
     "Twenty", "Thirty", "Forty", "Fifty",
     "Sixty", "Seventy", "Eighty", "Ninety",
     "Hundred", "Thousand", "Million"}

function wrd(a1, a2, a3)
  return (a1 + a2 + a3 > 0) and (((a1 > 0) and s[a1] .. s[28] or "") ..
                                 ((a2 > 1) and s[18 + a2] .. s[a3] or s[10*a2 + a3]))
end

for line in io.lines(arg[1]) do
  local b, c = tonumber(line), {}
  if b == 0 then
    io.write("Zero")
  else
    for i = 9, 1, -1 do
      c[i], b = b%10, div(b, 10)
    end
    local r = wrd(c[1], c[2], c[3])
    if r then io.write(r .. s[30]) end
    r = wrd(c[4], c[5], c[6])
    if r then io.write(r .. s[29]) end
    r = wrd(c[7], c[8], c[9])
    if r then io.write(r) end
  end
  io.write("Dollars\n")
end

File: hard/print_check.hs
import System.Environment (getArgs)

nstr :: [String]
nstr = ["Zero", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]

wrd :: [Int] -> String
wrd [0, x, y] = wrd [x, y]
wrd [0, x]    = wrd [x]
wrd [0]       = ""
wrd [x, y, z] = nstr!!x ++ "Hundred" ++ wrd [y, z]
wrd [x, y]    | x > 1     = nstr!!(x + 18) ++ wrd [y]
              | otherwise = nstr!!(10 + y)
wrd [x]       = nstr!!x

dolla                            :: [Int] -> String
dolla [0, 0, 0, x, y, z, a, b, c] = dolla [x, y, z, a, b, c]
dolla [0, 0, 0, x, y, z]          = dolla [x, y, z]
dolla [0, 0, 0]                   = ""
dolla [x, y, z, a, b, c, d, e, f] = wrd [x, y, z] ++ "Million" ++ dolla [a, b, c, d, e, f]
dolla [x, y, z, a, b, c]          = wrd [x, y, z] ++ "Thousand" ++ dolla [a, b, c]
dolla [x, y, z]                   = wrd [x, y, z]

splitDig  :: Int -> [Int]
splitDig x = [div x 100000000, mod (div x 10000000) 10, mod (div x 1000000) 10, mod (div x 100000) 10, mod (div x 10000) 10, mod (div x 1000) 10, mod (div x 100) 10, mod (div x 10) 10, mod x 10]

dollar  :: Int -> String
dollar x | x == 0    = "ZeroDollars"
         | otherwise = dolla (splitDig x) ++ "Dollars"

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map (dollar . read) $ lines input
